File: 144-preorder-traversal.js
/*
uh yeah it's a preorder tree traversal
*/

/*
recrusive solution
*/

var preorderTraversal = function(root) {
    let res = [];

    function preorder(node){
        if (!node) return;
        res.push(node.val);
        preorder(node.left);
        preorder(node.right);
    }

    preorder(root);

    return res;
};

File: 144-preorder-traversal.py
# see js file

class Solution:
    def preorderTraversal(self, root: Optional[TreeNode]) -> List[int]:
        res = []
        self.preorder(root, res)

        return res

    def preorder(self, node, res):
        if not node:
            return None
        res.append(node.val)
        self.preorder(node.left, res)
        self.preorder(node.right, res)
